### ISSUE TITLE
Add car to first pt station

### ISSUE THREAD

**Issue by subrina0013** [matsim-org/matsim-code-examples#694]
Hi.. Is there a way to add a car ride to the 1st pt station? If an agent has pt trip, all the transfers including the 1st trip from origin (even from home) is either walk or pt. So, is there a way to consider car parking at the pt stations? Also, is there a max distance defined that an agent will walk to get a pt? Can I set a max walking distance to the starting pt station so that agents live outside that walking distance will take car and maybe park at the stations?
Thank you so much

**Comment by Janekdererste** [comment]
Yes, in principle there is. @jakobrehmann might know how this works.

**Comment by jakobrehmann** [comment]
Hello and sorry for the late reply. Yes, car-based access/egress to pt stations is possible using the "SwissRailRaptor."

The following example should be a good starting point (you would replace taxi with car): https://github.com/matsim-org/matsim-libs/blob/master/contribs/av/src/main/java/org/matsim/contrib/av/intermodal/RunTaxiPTIntermodalExample.java

It is also possible to specify to what stations car-based access is allowed. This is done by adding a stopFilterAttribute & stopFilterValue to the paramSetTaxi. Then you have to modify the transit schedule to add the attributes to the stations that are supposed to be car-accessible. Does this help with you car parking question? As far as I know, we cannot set a parking capacity to limit how many cars can park at the station. 

Regarding max distance an agent will walk to a pt station: yes, this is implemented in the IntermodalAccessEgressParameterSet, as shown in the example. If you set both the InitialSearchRadius and the MaxSearchRadius to 1000, the agent will not walk more than 1km to get to a pt station.

**Comment by subrina0013** [comment]
Thank you so much. I will definitely try those and let you know.

**Comment by subrina0013** [comment]
Hi.. I just have a question. Does the distance of setInitialSearchRadius for walking mean the starting distance from origin to pt stations or it applies to all the walking between transits and transfers? 

If an agent takes pt and the 1st station is 2 km away from home, then he takes car to go to station right? But if two transit stations are more than 1 km away, then is he able to walk to transfer station? For example: there are several metro stations with heavily populated region that are located within 2 km distance but more than 1 km from each other. If an agent take transfer between those stations, what will happen?

**Comment by jakobrehmann** [comment]
Hello. I am not an expert in this code. However, as "setInitialSearchRadius" is in the IntermodalAccessEgressParameterSet, I think that this setting only applies to origin -> first pt station (access) or last pt station -> destination (egress) and not the transfers in between. I did some digging, and it seems that parameters relating to transfer-walks live in the [RaptorStaticConfig.java](https://github.com/matsim-org/matsim-libs/blob/master/matsim/src/main/java/ch/sbb/matsim/routing/pt/raptor/RaptorStaticConfig.java). Here is a selection (lines 36-41):

    ```private double beelineWalkConnectionDistance = 200.0;
    private double beelineWalkSpeed; // meter / second
    private double beelineWalkDistanceFactor = 1.0;

    private double minimalTransferTime = 60;
    private double transferWalkMargin = 5;```

The first parameter--beelineWalkConnectionDistance--represents the distance in meters that agents can walk to get from one stop to another stop of a nearby transit line.

**Comment by subrina0013** [comment]
Thank you so much for the clarification. 
However, I tried to run as in https://github.com/matsim-org/matsim-libs/blob/master/contribs/av/src/main/java/org/matsim/contrib/av/intermodal/RunTaxiPTIntermodalExample.java. But it kept running for days with no result. So, I guess I did some mistake somewhere. Here is the code. Could you please take a look?  
Thank you

[RunTest.txt](https://github.com/matsim-org/matsim-code-examples/files/8334156/RunTest.txt)

**Comment by jakobrehmann** [comment]
If you send me your log file & config, I can take a look.

**Comment by subrina0013** [comment]
Here are the files. 

[logfile.log](https://github.com/matsim-org/matsim-code-examples/files/8342817/logfile.log)
[RunTest.txt](https://github.com/matsim-org/matsim-code-examples/files/8342818/RunTest.txt)
[smartCity_march23_config.txt](https://github.com/matsim-org/matsim-code-examples/files/8342824/smartCity_march23_config.txt)

**Comment by jakobrehmann** [comment]
Looking at your log, it look 10 hours to **partially**  load your population. How many agents do you have? Did you sample the population to e.g. 1% or 10% of original size before use in MATSim?

**Comment by subrina0013** [comment]
I used a 10% sample of the whole population as the input population. My whole population was 3.2 million and I used 320,000 in the simulation. Still it didn’t finish loading.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by subrina0013** [comment]
> Looking at your log, it look 10 hours to **partially** load your population. How many agents do you have? Did you sample the population to e.g. 1% or 10% of original size before use in MATSim?

Is there any suggestion for me for 10% of the sample? How can I make it work?

**Comment by Janekdererste** [comment]
On the first sight your configuration looks alright. For starters your could try two things:
1. Throw RAM at your problem. Your computer seems to have 30GB of Memory.  You can give MATSim more by adding ```-Xmx26G``` to your runtime parameters (your command would then look soemething like this ```java -Xmx26G -jar your-matsim.jar path/to/your/config.xml```
2. Reduce the sample size of your population. Try to run things with a 1% or 0.1% Sample. Usually, having smaller sample sizes makes it easier to find errors because things load faster. Once you were able to run a small sample you can switch back to a bigger one. [Here](https://github.com/matsim-vsp/mosaik-2/blob/master/src/main/java/org/matsim/mosaik2/prepare/ReducePopulation.java) is a script to draw a smaller random sample of your current population

**Comment by FarnooshNm** [comment]
> Hello and sorry for the late reply. Yes, car-based access/egress to pt stations is possible using the "SwissRailRaptor."
> 
> The following example should be a good starting point (you would replace taxi with car): https://github.com/matsim-org/matsim-libs/blob/master/contribs/av/src/main/java/org/matsim/contrib/av/intermodal/RunTaxiPTIntermodalExample.java
> 
> It is also possible to specify to what stations car-based access is allowed. This is done by adding a stopFilterAttribute & stopFilterValue to the paramSetTaxi. Then you have to modify the transit schedule to add the attributes to the stations that are supposed to be car-accessible. Does this help with you car parking question? As far as I know, we cannot set a parking capacity to limit how many cars can park at the station.
> 
> Regarding max distance an agent will walk to a pt station: yes, this is implemented in the IntermodalAccessEgressParameterSet, as shown in the example. If you set both the InitialSearchRadius and the MaxSearchRadius to 1000, the agent will not walk more than 1km to get to a pt station.

@jakobrehmann 

Hi,

I'm trying to implement car-based access/egress. I modified the transit schedule XML file to add an attribute to stations that are car-accessible. However, when running MATSim, I get an error stating that the added attribute must be declared for the element type "stopFacility".

Is there a way to address this issue? Or should I define the attribute within the code instead of modifying the XML file? I'd greatly appreciate any guidance or examples of how this has been implemented in code.

Thank you.

**Comment by FarnooshNm** [comment]
I think I figured out how to add the attribute in the code. This is what I've added to enable park and ride for four transit stops:

        TransitSchedule transitSchedule = scenario.getTransitSchedule();
        Set<Id<TransitStopFacility>> carAccessibleStopIds = Set.of(
                Id.create("G31S.link:pt_G31S", TransitStopFacility.class),
                Id.create("G31N.link:pt_G31N", TransitStopFacility.class),
                Id.create("F22S.link:pt_F22S", TransitStopFacility.class),
                Id.create("F22N.link:pt_F22N", TransitStopFacility.class)
        );

        
        for (TransitStopFacility stopFacility: transitSchedule.getFacilities().values()){
            if (carAccessibleStopIds.contains(stopFacility.getId())) {
                stopFacility.getAttributes().putAttribute("parkAndRide", "true");
            }
        }

However, I'm getting another error now:

> org.matsim.core.mobsim.qsim.pt.TransitQSimEngine$TransitAgentTriesToTeleportException: Agent 258919_1 tries to enter a transit stop at link pt_G31S but really is at 39612790_2!

The car and transit networks I'm using are separate and not connected. My guess is that this is causing the issue. The transit stops I’ve defined as car-accessible are not technically car-accessible because they are not connected to the car network. Since I’m defining walk and car as access and egress modes, I expect the car to get close to the car-accessible station and then reach it via the walk access mode. However, I’m not sure how to properly implement this. I would appreciate any help.

**Comment by FarnooshNm** [comment]
In case anyone runs into the same issue, I ended up adding dummy links (allowing car access) to connect the road network to the transit network (car-accessible stops), which seems to have resolved the error.
However, after adding "car" as an access/egress mode, I’m seeing many unfinished transit trips. Something seems to be off, but I’m not sure what the root cause is.

**Comment by paulheinr** [comment]
> I ended up adding dummy links (allowing car access) to connect the road network to the transit network (car-accessible stops)

This sounds a bit hacky. Normally, the PT network is separated from the car network because it is on another abstraction level. I am not sure, what other (unwanted) consequences this has. 

As discussed in #722 implementing different access egress modes is not easy. I think using the SwissRailRaptor as @jakobrehmann suggested is the cleanest way. However, I am not an expert for this.
